### PR TITLE
Enable speculating on expressions with expression variables

### LINKED ISSUE #14384: Speculatively binding an invocation that contains a "DeclarationExpression" crashes the compiler
As part of a new feature i'm writing, i rewrite an existing invocation to use a DeclarationExpression.  i.e. i transform code for the form:

```Foo(out i)``` into ```Foo(out int i)```.

To ensure i don't break semantics i speculatively bind the new expression to see if it binds to the same method symbol as before.  Currently, this crashes the compiler here (stack is below):

```c#
            // Is this a field?
            GlobalExpressionVariable expressionVariableField = LookupDeclaredField(declarationExpression.VariableDesignation());

            if ((object)expressionVariableField == null)
            {
                // We should have the right binder in the chain, cannot continue otherwise.
                throw ExceptionUtilities.Unreachable;
            }
```

The reason seems to be that the we fail the code above this.  Namely: 

```c#
            // Is this a local?
            SourceLocalSymbol localSymbol = this.LookupLocal(declarationExpression.Identifier());

            if ((object)localSymbol != null)
```

Because this out-variable is not in the locals map, we don't find the local for it and we call into the code that crashes.

Now, i'm not sure what the right thing is here.  It's possible that GetSpeculativeSymbolInfo just can't handle expressions that introduce variables.  If so, that's fine (i'll find another way to verify that semantics don't change).  However, we should not crash when perfectly reasonable inputs as passed in.  So, at the very least we need to harden ourselves to this code.  

That said, i woudl very much like it if this scenario actually worked :)

<details>
```
Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindOutVariableArgument(Microsoft.CodeAnalysis.CSharp.Syntax.DeclarationExpressionSyntax declarationExpression, Microsoft.CodeAnalysis.DiagnosticBag diagnostics) Line 2102	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindArgumentValue(Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentSyntax argumentSyntax, bool allowArglist, Microsoft.CodeAnalysis.RefKind refKind) Line 2090	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindArgumentAndName(Microsoft.CodeAnalysis.CSharp.AnalyzedArguments result, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, bool hadError, Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentSyntax argumentSyntax, bool allowArglist, bool isDelegateCreation) Line 2060	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindArgumentsAndNames(Microsoft.CodeAnalysis.SeparatedSyntaxList<Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentSyntax> arguments, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments result, bool allowArglist, bool isDelegateCreation) Line 2015	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindArgumentsAndNames(Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentListSyntax argumentListOpt, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments result, bool allowArglist, bool isDelegateCreation) Line 1988	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindInvocationExpression(Microsoft.CodeAnalysis.CSharp.Syntax.InvocationExpressionSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics) Line 153	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, bool invoked, bool indexed) Line 452	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, bool invoked, bool indexed) Line 396	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax node, Microsoft.CodeAnalysis.DiagnosticBag diagnostics) Line 391	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSpeculativelyBoundExpressionHelper(Microsoft.CodeAnalysis.CSharp.Binder binder, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax expression, Microsoft.CodeAnalysis.SpeculativeBindingOption bindingOption, Microsoft.CodeAnalysis.DiagnosticBag diagnostics) Line 250	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSpeculativelyBoundExpression(int position, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax expression, Microsoft.CodeAnalysis.SpeculativeBindingOption bindingOption, out Microsoft.CodeAnalysis.CSharp.Binder binder, out System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.CSharp.Symbol> crefSymbols) Line 310	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetSpeculativeSymbolInfo(int position, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax expression, Microsoft.CodeAnalysis.SpeculativeBindingOption bindingOption) Line 680	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.CSharpExtensions.GetSpeculativeSymbolInfo(Microsoft.CodeAnalysis.SemanticModel semanticModel, int position, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax expression, Microsoft.CodeAnalysis.SpeculativeBindingOption bindingOption) Line 533	C#
 	Microsoft.CodeAnalysis.CSharp.Features.dll!Microsoft.CodeAnalysis.CSharp.InlineDeclaration.CSharpInlineDeclarationCodeFixProvider.SemanticsChanged(Microsoft.CodeAnalysis.SemanticModel semanticModel, Microsoft.CodeAnalysis.CSharp.Syntax.VariableDeclarationSyntax declaration, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax invocationOrCreation, Microsoft.CodeAnalysis.CSharp.Syntax.TypeSyntax newType, Microsoft.CodeAnalysis.CSharp.Syntax.IdentifierNameSyntax identifier, Microsoft.CodeAnalysis.CSharp.Syntax.DeclarationExpressionSyntax declarationExpression) Line 130	C#
```
</details>

### PR DESCRIPTION
**Pull request by gafter**
Fixes #14384 

@dotnet/roslyn-compiler Please review.

### CODE REVIEW

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Please review this test-only addition.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Please review. The crash has been reproduced and fixed in this revision.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 1]
```diff
@@ -215,10 +215,10 @@ internal Binder GetSpeculativeBinder(int position, ExpressionSyntax expression,
             }
 
             // May be binding an expression in a context that doesn't have a LocalScopeBinder in the chain.
-            return new LocalScopeBinder(binder);
+            return new ExpressionVariableBinder(expression, binder);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: >ExpressionVariableBinder [](start = 23, length = 24)

I think the code will be more robust long term if we will not create ExpressionVariableBinder explicitly, but instead always create an ExecutableCodeBinder and then request a binder for the root from it.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 2]
```diff
@@ -229,7 +229,7 @@ private Binder GetSpeculativeBinderForAttribute(int position)
             }
 
             // May be binding an expression in a context that doesn't have a LocalScopeBinder in the chain.
-            return new LocalScopeBinder(binder);
+            return new ExpressionVariableBinder(attribute, binder);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: >return new ExpressionVariableBinder(attribute, binder); [](start = 12, length = 55)

Similar suggestion.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: How would you suggest I test for the difference between your suggestion and the implementation here?

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Same question here

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: At the moment, there is no way to test for an observable difference. However, once we add "match" expression, or an alternative, that would introduce a scope within a portion of an expression, creating ExpressionVariableBinder for the top level scope wouldn't be sufficient. ExecutableCodeBinder, however, will be able to deal with this situation.

**Comment by gafter** [comment]
I am not making a corresponding change in `SourceFixedFieldSymbol` as it breaks a couple of tests, and therefore requires further investigation. I'll file a separate issue for that.

**Comment by AlekseyTs** [comment]
> I'll file a separate issue for that.

Sounds good. Thanks for trying it out!

**Comment by AlekseyTs** [comment]
LGTM

**Comment by gafter** [comment]
@agocke Please review.

**Comment by AlekseyTs** [comment]
> I am not making a corresponding change in SourceFixedFieldSymbol as it breaks a couple of tests, and therefore requires further investigation.

I think we shouldn't have parented that binder to the field symbol, the containing member from the parent binder should have been used instead.

**Comment by gafter** [comment]
I've added a link to that comment in #14510

**Comment by gafter** [comment]
@MattGertz tagging for RC. This fixes a crash in IDE scenarios.

**Pull request merged**